include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,367 @@ enum E_DESCRIPTOR_TYPE : uint8_t
-                    --immSamplersOffset;//going back to prev state
-					for (uint32_t s = 0ull; s < bnd_in.count; ++s)
-						m_samplers->operator[](immSamplersOffset+s) = bnd_in.samplers[s];
-					immSamplersOffset += bnd_in.count;
-				}
-			}
+	bool isIdenticallyDefined(const IDescriptorSetLayout<sampler_type>* _other) const
+	{
+		if (!_other || getTotalBindingCount() != _other->getTotalBindingCount())
+			return false;
 
-			std::fill_n(m_descriptorCount, EDT_COUNT, 0ull);
-            if (m_bindings)
-            {
-                for (size_t i = 0ull; i < m_bindings->size(); ++i)
-                {
-                    auto& bnd = m_bindings->operator[](i);
+		auto areRedirectsEqual = [](const CBindingRedirect& lhs, const CBindingRedirect& rhs) -> bool
+		{
+			const auto memSize = lhs.getRequiredMemorySize();
+			if (memSize != rhs.getRequiredMemorySize())
+				return false;
 
-                    static_assert(sizeof(size_t) == sizeof(bnd.samplers), "Bad reinterpret_cast!");
-                    if (bnd.type == EDT_COMBINED_IMAGE_SAMPLER && bnd.samplers)
-                        bnd.samplers = m_samplers->data() + reinterpret_cast<size_t>(bnd.samplers) - 1ull;
+			if (std::memcmp(lhs.m_data.get(), rhs.m_data.get(), memSize) != 0)
+				return false;
 
-					m_descriptorCount[bnd.type] += bnd.count;
-                }
+			return true;
+		};
 
-                // TODO: check for overlapping bindings (bad `SBinding` definitions)
-                std::sort(m_bindings->begin(), m_bindings->end());
-            }
+		for (uint32_t t = 0u; t < static_cast<uint32_t>(IDescriptor::E_TYPE::ET_COUNT); ++t)
+		{
+			if (!areRedirectsEqual(m_descriptorRedirects[t], _other->m_descriptorRedirects[t]))
+				return false;
 		}
-		virtual ~IDescriptorSetLayout() = default;
 
-		core::smart_refctd_dynamic_array<SBinding> m_bindings;
-		core::smart_refctd_dynamic_array<core::smart_refctd_ptr<sampler_type>> m_samplers;
+		if (!areRedirectsEqual(m_immutableSamplerRedirect, _other->m_immutableSamplerRedirect))
+			return false;
 
-	public:
-		bool isIdenticallyDefined(const IDescriptorSetLayout<sampler_type>* _other) const
+		if (!areRedirectsEqual(m_mutableSamplerRedirect, _other->m_mutableSamplerRedirect))
+			return false;
+
+		if (!m_samplers && !_other->m_samplers)
 		{
-			if (!_other || getBindings().size()!=_other->getBindings().size())
+			return true;
+		}
+		else if (!m_samplers || !_other->m_samplers)
+		{
+			return false;
+		}
+		else
+		{
+			const auto samplerCount = m_samplers->size();
+			if (samplerCount != _other->m_samplers->size())
 				return false;
 
-			const size_t cnt = getBindings().size();
-			const SBinding* lhs = getBindings().begin();
-			const SBinding* rhs = _other->getBindings().begin();
-			for (size_t i = 0ull; i < cnt; ++i)
-				if (lhs[i] != rhs[i])
+			for (uint32_t i = 0u; i < samplerCount; ++i)
+			{
+				if (m_samplers->begin()[i] != _other->m_samplers->begin()[i])
 					return false;
+			}
 			return true;
 		}
```

Comment:
`std::equal` would be nice cause it takes care of everything in this entire else clause
```cpp
template< class InputIt1, class InputIt2 >

constexpr bool equal( InputIt1 first1, InputIt1 last1,
                      InputIt2 first2, InputIt2 last2 );
```

you can simply do
```cpp
if (m_samplers && _other->m_samplers)
    return std::equal(m_samplers->begin(),m_samples->end(),_other->m_samplers->begin(),_other->m_samples->end());
else
   return !m_samplers && !_other->m_samplers;
```

Reply:
Resolved.